### PR TITLE
Invalidate tokens. Expire Tokens. And config options.

### PR DESCRIPTION
**Pull request by scottrobertson**
# Expire tokens

We just use the built in JWT expire system. This length of time can be configured with `Tokenable::Config.lifespan`. Default is forever.

# Config options

```ruby
Tokenable::Config.lifespan
Tokenable::Config.secret
```

# Invalidate Tokens

We keep track of a `tokenable_verifier" in the DB and also in the JWT token.

If the verifier is changed/removed from the DB, and a user tries to use the JWT token, then the request will be denied.

Also adding messages to a lot of exceptions.

### CODE REVIEW

File: lib/tokenable/revokeable.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+module Tokenable
+  module Revokable
```
Comment (marcqualie) [inline lib/tokenable/revokeable.rb:2]:
```suggestion
  module Revocable
```

File: lib/tokenable/revokeable.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    def revoke_tokens_hash
+      raise Tokenable::Unauthorized.new("#{revoke_tokens_key} field is missing") unless self.has_attribute?(revoke_tokens_key)
+
+      read_attribute(revoke_tokens_key).to_h
+    end
+
+    # TODO: make a config option
+    def revoke_tokens_key
+      :revoke_tokens
```
Comment (marcqualie) [inline lib/tokenable/revokeable.rb:30]: They're not technically revoke tokens, they're salts/secrets that are used to check for validity. The exat opposite of a revoke token in terms of terminology.

We could instead use something like: `tokenable_secret_keys` `tokenable_jwt_verifiers`

File: lib/tokenable/revokeable.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+module Tokenable
+  module Revokable
+    extend ActiveSupport::Concern
+
+    def token_revoked?(revoke_token)
```
Comment (marcqualie) [inline lib/tokenable/revokeable.rb:5]: You're not actually checking if the token is reokved, the exact opposite. Rather, we want to check if the `secret` is valid.

File: lib/tokenable/authable.rb [hunk 4]
```diff
@@ -35,30 +36,35 @@ def user_class
...
+      if user.respond_to?(:issue_revoke_token)
+        jwt_data[:revoke_token] = user.issue_revoke_token
+      end
+
+      JWT.encode(jwt_data, jwt_secret, 'HS256')
     end
 
     def jwt_user_id
-      jwt['data']['user_id']
+      jwt['user_id']
```
Comment (marcqualie) [inline lib/tokenable/authable.rb:55]: Much better having it in root 👍

File: lib/tokenable/revokeable.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      !revoke_token.in?(revoke_tokens_hash.keys)
+    end
+
+    def issue_revoke_token
+      issued_token = SecureRandom.hex
+      self.update(
+        revoke_tokens_key => revoke_tokens_hash.merge(
```
Comment (marcqualie) [inline lib/tokenable/revokeable.rb:12]: We probably don't want to add a token every time, thish would grow huge and become a performace bottleneck in rails and the DB.

We should instead pick the most recent one each time a new jwt token is issued. 99% of the time, there will be a single token in use. The second one comes into play when you cycle the tokens but want to keep existing clients working while you update them. Think like aws credentials, they only let you have two as they're speficially designed for cycling every X days.

**Review by marcqualie — CHANGES_REQUESTED** [review]

File: lib/tokenable/revokeable.rb (hunk 5 above)
Comment (scottrobertson) [reply lib/tokenable/revokeable.rb:12]: This is what makes me think we should just go with a single "secret" for now. It removes a lot of complexity. We would just keep it the same until it's under.

File: lib/tokenable/authable.rb (hunk 4 above)
Comment (scottrobertson) [reply lib/tokenable/authable.rb:55]: @marcqualie we will have to switch back to `data` when we add expiry i think.

File: README.md [hunk 6]
```diff
@@ -47,12 +47,51 @@ You can also create your own stragery. This is as simple as creating a method on
...
+```ruby
+class User < ApplicationRecord
+  include Tokenable::Verifier
+end
+```
+
+And running the following migration:
+
+```bash
+rails g migration AddTokenableVerifierToUsers tokenable_verifier:uuid
```
Comment (marcqualie) [inline README.md:69]: `uuid` is a bit strict. We want to allow longer keys, and just alphanumeric ideally.

Plus in future we can support multiple keys by just splitting on `,` for backwards compatibility and ability to cycle. We can't do that with a `uuid` field on the db.

**Review by marcqualie — CHANGES_REQUESTED** [review]

**Review by marcqualie — APPROVED** [review]

**Pull request merged**